### PR TITLE
Nearest neighbors

### PR DESCRIPTION
**Pull request by gwenwhite**
This is still a WIP but think I calculated nearest neighbors. I am a bit confused on the "average nearest neighbors" part of this. Will I be finding the average nearest neighbors for a random particle in the system?

### CODE REVIEW

File: MCMC/utils.py [hunk 1]
```diff
@@ -2,6 +2,33 @@
 import numpy as np
 from numba import jit
 
+def avg_nn(sim_obj, n_frames, r_max=3):
+    box = [sim_obj.L, sim_obj.L,0]
+    frame_avg_nn = []
+    for points in sim_obj.system_history[-n_frames:]:
+        points = np.append(
+            points, np.zeros((points.shape[0], 1)),axis=1
```
Comment (madilynpaul) [inline MCMC/utils.py:10]: What is the np.zeros((points.shape[0],1)) doing?

File: MCMC/utils.py [hunk 2]
```diff
@@ -2,6 +2,33 @@
...
+        )
+        aq = freud.locality.AABBQuery(box, points)
+        nlist = aq.query(points, {'r_max': 3}).toNeighborList()
+        neighbors = []
```
Comment (madilynpaul) [inline MCMC/utils.py:14]: What does the aq function/variable do?

**Review by madilynpaul — APPROVED** [review]
I think that everything looks good. I left a couple questions on stuff that I just didn't understand because I've never used those functions.

**Comment by marjanAlbouye** [comment]
> I have the code calculating the average nearest neighbors for every frame but I am wondering if that is necessary to do for every frame or not? The frame number doesn't seem to effect the number of average nearest neighbors.

I think keeping it as it is right now should be fine. The decision about calculating the avg nn on a bunch of frames or just the last frame of the simulation depends on what kind of analysis we want to do later. That could be a good question/topic to discuss during the meeting.

**Review by marjanAlbouye — APPROVED** [review]
LGTM!

File: MCMC/utils.py (hunk 2 above)
Comment (chrisjonesBSU) [reply MCMC/utils.py:14]: This is a function in the freud analysis package we use a lot. Here it is essentially being used to search through the particles in the system (points and box) to find neighbors, or neighborhoods of particles.  The docs for freud have a pretty good overview of it [here](https://freud.readthedocs.io/en/latest/gettingstarted/tutorial/neighborfinding.html#neighbors)

File: MCMC/utils.py (hunk 1 above)
Comment (chrisjonesBSU) [reply MCMC/utils.py:10]: Our system is 2D, which means all of our coordinates are only (x, y). Sometimes we need 3 dimensional coordinates to use other tools, like freud, or writing to a gsd file. This line adds a z coordinate of 0 to all of our 2D coordinates.

**Comment by chrisjonesBSU** [comment]
This looks good to me. @gwenwhite I kinda of led you astray last week when helping you with the RDF function. We made the mistake of writing these analysis functions before determining how data would be logged. So, last week it made sense (to me at least) to write these functions in a way that pull the needed information directly from the `Simulation()` object. 

Now that we have the logging process solidified, and are logging positions to GSD files, we'll have to re-work the RDF, S(q) and NN functions to be compatible with GSDs.  The good news is that the basic framework for these will stay the same.  If you'd like, I can make these changes myself, so you can move on to other things.

**Review by chrisjonesBSU — APPROVED** [review]
Let's merge these now, and make the changes for GSD compatibility in a separate PR.

**Pull request merged**